fix: remove unused components transitively

fb55 opened the pull request:
## What/Why/How?

Removes components if their parent paths were removed. Loops until no more changes were made.

## Reference

Fixes https://github.com/Redocly/redocly-cli/issues/874

## Testing

Tests were updated to cover the new logic.

## Check yourself

- [ ] Code changed? - Tested with redoc/reference-docs/workflows (internal)
- [x] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

fb55 commented:
@tatomyr Please have a look!

tatomyr commented:
@fb55 sure thing, will take a look!

tatomyr reviewed (changes requested):
Sorry for the delay! 

Although I like the approach, there's one major issue with that solution to be fixed. 
Also, I left a few style change suggestions for the sake of readability (also applicable to the OAS2 variant). 

And one more thing. Whatever changes here should also apply to the corresponding rule [no-unused-components](https://github.com/Redocly/redocly-cli/blob/main/packages/core/src/rules/oas3/no-unused-components.ts). 
 
Again, thank you for your efforts. I hope you'll find my comments useful 🙂.

fb55 commented on packages/core/src/decorators/oas3/remove-unused-components.ts:
```diff
@@ -49,17 +56,32 @@ export const RemoveUnusedComponents: Oas3Decorator = () => {
+          for (const [path, { used, name, componentType }] of components) {
+            const isUsed = used?.some(
+              (location) =>
+                !removedPaths.some((removed) => location.absolutePointer.startsWith(removed))
```
Great catch! Adding a check for a trailing slash.

fb55 commented:
Thanks for the review, addressed all comments!

> And one more thing. Whatever changes here should also apply to the corresponding rule [no-unused-components](https://github.com/Redocly/redocly-cli/blob/main/packages/core/src/rules/oas3/no-unused-components.ts).

Do you mean introducing the additional data structures for tracking where a component is used?

tatomyr commented:
> Do you mean introducing the additional data structures for tracking where a component is used?

Actually, at a second glance, I'm not sure we should apply the same logic in the rule itself... We're not removing transitive components but rather checking the current state. 

Let me have another look.

tatomyr commented on packages/core/src/decorators/oas2/remove-unused-components.ts:
```diff
@@ -16,15 +16,44 @@ export const RemoveUnusedComponents: Oas2Decorator = () => {
+        (location) =>
+          !removedPaths.some(
+            (removed) =>
+              location.absolutePointer.startsWith(removed) &&
+              (location.absolutePointer.length === removed.length ||
```
Isn't this the same as the following?
```
location.absolutePointer === removed ||
location.absolutePointer.startsWith(removed + '/') 

```

fb55 replied:
Yes, but it isn't comparing the strings twice & doesn't create a string copy. We are running through this loop a lot, so making it performant seems important.

fb55 replied:
Added a comment!

tatomyr commented on packages/core/src/decorators/oas2/remove-unused-components.ts:
```diff
@@ -16,15 +16,44 @@ export const RemoveUnusedComponents: Oas2Decorator = () => {
+    const removedLengthStart = removedPaths.length;
+
+    for (const [path, { usedIn, name, componentType }] of components) {
+      const used = usedIn?.some(
```
```suggestion
      const used = usedIn.some(
```

Is the optional chaining still needed?

fb55 replied:
It is not!

tatomyr reviewed (changes requested):
WIth the latest changes it doesn't seem to remove transitive references anymore 😅.

fb55 commented:
> WIth the latest changes it doesn't seem to remove transitive references anymore 😅

Tests are passing, and the codebase I am looking at (~1000 auto-generated components with ~20% unused) is not showing any unused component errors after processing. But the reported number of removed components has gone down from ~220 to ~160 — not quite sure about what happened here.

tatomyr commented:
> Tests are passing

Indeed they are. Sorry, mixed up with another errors 🙂.

tatomyr commented on packages/core/src/decorators/oas3/__tests__/remove-unused-components.test.ts:
```diff
@@ -168,4 +168,146 @@ describe('oas3 remove-unused-components', () => {
+              properties:
+                link:
+                  $ref: '#/components/schemas/Unused'
+            Used: # <-- this will be removed INCORRECTLY because the pointer starts with a previously removed path 
```
Could you change the comment? It *will not* be removed after the fix 🙂.

tatomyr commented on packages/core/src/decorators/oas2/__tests__/remove-unused-components.test.ts:
```diff
@@ -152,4 +152,77 @@ describe('oas2 remove-unused-components', () => {
+              summary: List all pets
+        definitions:
+          Unused:
+            enum:
+              - 1
+              - 2
```
Could you also clean up the tests from the needless fields like that one?

fb55 replied:
This was copied from other test cases, and I am hesitant to include a larger cleanup here.

fb55 commented:
@tatomyr Anything I can help with to make this move forward?

fb55 commented:
Realised there was a comment I had not addressed, just did so!

lornajane reviewed (approved):
Changelog is good (approval from @Redocly/technical-writers)

tatomyr reviewed (approved):
Great job @fb55! 🔥 
Your changes will be available in the next release.

fb55 commented:
@tatomyr Thanks for shepherding this to the finish line!
